### DISCUSSION TITLE
Substring Query to filter data

### DISCUSSION THREAD

**Discussion by ashokgudur** [ravendb/ravendb#13930] [General]
Hi,

<img width="400" alt="Screenshot 2022-03-29 111240" src="https://user-images.githubusercontent.com/3616094/160541554-2deed07e-8b2e-4696-9459-758d7d527032.png">

I have a document with identities generated with pipe as given above.
I want to read docs by filtering on number part of the id like...

session.Query<Person>()
            .Where(x => Convert.ToInt32(x.Id.Substrnig(1)) >= 2000)
            .Where(x => Convert.ToInt32(x.Id.Substrnig(1)) <= 1500)
            .ToList();

But RavenDB says its not supported

thanks.

**Comment by krivulcik** [comment]
It's impossible to query on computed data - it's not present in the index, so it wouldn't be performant enough.

You can achieve this in two ways:
 - Create an extra field in the index where the computation would be done (i.e., the field would be `IdNumPart = Convert.ToInt32(x.Id.Substring(2))` (2 to skip the slash)). This requires you to have a static index. I'm not sure if you already have one.
 - Store the value in an extra field in the object, allowing you to query on auto indexes. However, this requires extra storage and causes complications if you need to change the document ID (you need to change the derived value).

**Reply by ashokgudur** [reply]
>>>Store the value in an extra field in the object, allowing you to query on auto indexes. However, this requires extra storage and causes complications if you need to change the document ID (you need to change the derived value).

The Id is server side generated value. I am not sure how to capture the value before or after it is generated/saved and put in an extra field. I will not know what value the server has generated? Would the after-save-changes-event gives any hints on what's the new Id is generated on saving? 
I need to use patching? call save two times?

thanks

**Reply by krivulcik** [reply]
Well, I'm not sure about that. I wouldn't use this approach due to the disadvantages I mentioned. I included it only to show more options.

**Comment by ashokgudur** [comment]
Ok. It makes sense.
I will try using the static index.
thanks.

**Reply by ayende** [reply]
Can you explain the scenario where you want to do range queries on the numbers? Are you trying to do paging?

**Reply by ashokgudur** [reply]
No not paging. 
I have an actionlog table/document. In that I have a field to hold the log numbers. This number should be a server-side globally autoincremented field. Since we don't have a direct auto generated (increment) feature, I'm manually incrementing the field myself via code by finding the last/max number etc. However that is making duplicates when concurrent requests are made by the users.

I saw that I can use autogenerated Ids with pipe | char which provides this feature. I would actually need to query the log by a rage to fetch the entries for a userId and for the range. But the Id field is string. I saw there is extra param provided in query.OrderBy to treat strings as numbers but I cannot use it in filtering. Hence I need to use the substring and convert the value into number and filter it.

thanks.

**Reply by ayende** [reply]
You can use identities from your code directly

`long id = store.Maintenance.Send(new NextIdentityForOperation("users"));`

That would give you the same as `users|` , basically, but usable ahead of time.

**Reply by ashokgudur** [reply]
That works. 

I noticed there are unused Identities/names/prefixes as in the image

<img width="758" alt="Screenshot 2022-03-30 192713" src="https://user-images.githubusercontent.com/3616094/160852132-9d925e00-1a3c-47ae-a87b-c7c82ded795f.png">


How remove them? They are all unused. Testing prefixex


Thanks,

**Reply by ashokgudur** [reply]
Is the NextIdentityForOperation
self increasing? 
I mean calling this will increase the current value globally visible to other sessions of other store objects created elsewhere to prevent duplicates?

**Reply by garayx** [reply]
Hi

yes, calling `NextIdentityForOperation` make a server call and returns you an exclusive value for this identity, calling the this operation elsewhere will return you another exclusive value, you can see more info on this operation here:
https://ravendb.net/docs/article-page/5.3/csharp/client-api/document-identifiers/working-with-document-identifiers#using-operations
